### PR TITLE
add handle_unknown function to sentiment.jl

### PR DESCRIPTION
**Pull request by aquatiko**
Added a default function to handle unknown words while using `SentimentAnalyser`.
Default function: Skips the word

Closes: https://github.com/JuliaText/TextAnalysis.jl/issues/83
Closes: https://github.com/JuliaText/TextAnalysis.jl/issues/122

### CODE REVIEW

**Comment by aquatiko** [comment]
This got me thinking, here we have just added a default function to skip the unknown words from the dictionary. But if the user is giving his own function, then the result should be some `word_id` from the dictionary. This, is because the variable `res` holds the ids of all words, which is the inputted to the model to operate and later find an embedding for it.
Maybe, we should add  a step to first process the unknown word by the user given function and again check for it's presence in dictionary and if still not present, then we can safely ignore it.
Suggestions @aviks?

**Comment by ksteimel** [comment]
@aquatiko What do you get if you run the following under your additions?

``` 
using TextAnalysis
model = SentimentAnalyzer()
doc = Document("some sense duh")
model(doc)
```

I am getting an out of index error for any document that includes the word "duh" after implementing your proposed changes in my local copy of TextAnalysis. I suspect it's something I didn't do correctly while emulating what you did in this pr.

**Comment by aquatiko** [comment]
@ksteimel This was one interesting bug to figure out!!
It seems like the model was trained on 8k+ words but the no. of unique words which actually have embedding is 5k.
Thus the word like `duh` or `rohit` were actually present in the dictionary but didn't have an embedding.
Thus the implementation became a bit buggy.
I added some condition, It seems to be working just fine now.

File: src/sentiment.jl [hunk 1]
```diff
@@ -35,7 +35,7 @@ function flatten(x)
     return reshape(x, (l, 1))
 end
 
-function get_sentiment(ip::Array{T, 1}, weight, rwi) where T <: AbstractString
+function get_sentiment(ip::Array{T, 1}, weight, rwi, handle_unknown) where T <: AbstractString
```
Comment (oxinabox) [inline src/sentiment.jl]:
```suggestion
function get_sentiment(ip::AbstractVector{<:AbstractString}, weight, rwi, handle_unknown)
```

also `handle_unknown` as a function argument should be first.
https://docs.julialang.org/en/v1/manual/style-guide/index.html#Write-functions-with-argument-ordering-similar-to-Julia-Base-1

File: src/sentiment.jl [hunk 2]
```diff
@@ -45,7 +45,11 @@ function get_sentiment(ip::Array{T, 1}, weight, rwi) where T <: AbstractString
     end
     res = Array{Int, 1}()
     for ele in ip
-        push!(res, rwi[ele])
+	if ele in keys(rwi) && rwi[ele] <= 5000    # there are only 5000 unique embeddings
```
Comment (oxinabox) [inline src/sentiment.jl]: read that `5000` out of  the size of `a_1`, rather than hard-coding it?

File: src/sentiment.jl [hunk 3]
```diff
@@ -74,10 +78,22 @@ function Base.show(io::IO, s::SentimentAnalyzer)
 end
 
 
-function(m::SentimentModel)(text::Array{T, 1}) where T <: AbstractString
-    return get_sentiment(text, m.weight, m.words)
+function(m::SentimentModel)(text::Array{T, 1}, handle_unknown) where T <: AbstractString
```
Comment (oxinabox) [inline src/sentiment.jl]:
```suggestion
function (m::SentimentModel)(text::AbstractVector{<: AbstractString}, handle_unknown)
```

Also again `handle_unknown` should be first argument.

File: src/sentiment.jl [hunk 4]
```diff
@@ -74,10 +78,22 @@ function Base.show(io::IO, s::SentimentAnalyzer)
...
+ model = SentimentAnalyzer(doc)
+ model = SentimentAnalyzer(doc, handle_unknown)
+ ```
+ Return sentiment of the input doc in range 0 to 1, 0 being least sentiment score and 1 being
+ the highest:
+  -  doc              = Input Document for calculating document (AbstractDocument type)
+  -  handle_unknown   = A function for handling unknown words. Should return an array (default (x)->[])
+ """
+
+function(m::SentimentAnalyzer)(d::AbstractDocument, handle_unknown::Function = (x)->[])
```
Comment (oxinabox) [inline src/sentiment.jl]:
```suggestion
function (m::SentimentAnalyzer)(d::AbstractDocument, handle_unknown = x->tuple())
```

Also again `handle_unknown` should be first argument.

don't constrain callables to be subtypes of functions unless dispatching on them,
and it is maginally faster to use `tuple()` than `[]`

**Comment by aquatiko** [comment]
> @aviks Yeah, we can pass the dictionary to the `handle_unknown` function and the user can use the dictionary to modify his words, maybe apply a similarity match with his own corpus or some other similar method of choice. But, one thing is for sure, we have to drop the words which couldn't be brought into our dictionary's domain in regards to current model.

@oxinabox I have added the suggested changes but for adding more tests we have to decide if this should be the desired behaviour

**Comment by aquatiko** [comment]
Ping @aviks. Could we have some conclusions regarding this desired behaviour?

**Comment by aviks** [comment]
Yeah, this looks like a sensible strategy to me.

**Comment by aquatiko** [comment]
I have added the required changes but can't seem to find why some of `Platform: x86` tests fails :disappointed:

**Comment by oxinabox** [comment]
You are probably on top of this but

Still need to add:
 - Tests where `handle_unknown` throws an error
 - Tests were `handle_unknown` returns a word that is still unknown

**Comment by aquatiko** [comment]
These respective tests covers this:
>     * Tests where `handle_unknown` throws an error
>     * Tests where `handle_unknown` returns `[]`
https://github.com/JuliaText/TextAnalysis.jl/blob/d612724860ac51e30f9c8e626dacd520660c3251/test/sentiment.jl#L17
and
https://github.com/JuliaText/TextAnalysis.jl/blob/d612724860ac51e30f9c8e626dacd520660c3251/test/sentiment.jl#L12



>     * Tests where `handle_unknown` returns a word that is still unknown
https://github.com/JuliaText/TextAnalysis.jl/blob/d612724860ac51e30f9c8e626dacd520660c3251/test/sentiment.jl#L27

File: test/sentiment.jl [hunk 5]
```diff
@@ -8,4 +8,25 @@
...
+    # testing user given handle_unknown function
+    d = Document("a Horrible thing that Everyone Hates")
+   
+    @test m(d, (x) -> [lowercase(x)]) < 0.5
+
+    # when handle unknown returns a word which is still unknown
+    d = Document("some sense and some Hectic")
+   
+    @test m(d, (x) -> [lowercase(x)]) < 0.5
+
```
Comment (oxinabox) [inline test/sentiment.jl]:
```suggestion
    # Make it throw an error when when unknown word encountered
    d = Document("some sense and some Hectic")

    @test_throws ErrorException m(d, (x) -> error("OOV word $x encountered")
```

Comment (aquatiko) [reply test/sentiment.jl]: @oxinabox  Done!!

**Comment by aquatiko** [comment]
@aviks Would this need more work or is this good enough to be merged now?

**Pull request merged**